File: Sources/Views/Sidebar/GitSidebarViewModel.swift
// MARK: - GitSidebarViewModel
// Observable view model encapsulating all git-related state and operations
// for the sidebar git panel.
// macOS 14+, Swift 5.10

import AppKit
import Foundation
import OSLog

// MARK: - GitSidebarViewModel

/// Manages git state for the sidebar: statuses, branches, commit input,
/// push/pull operations, and AI commit message generation.
///
/// Extracted from `SidebarView` to reduce its complexity and separate
/// business logic from the view layer.
@Observable
@MainActor
final class GitSidebarViewModel {

    // MARK: - Dependencies

    private let gitService: any GitServicing
    private let aiCommitService: any AICommitServicing

    // MARK: - Git Multi-Project State

    /// Expanded project rows in the git section.
    var gitExpandedProjects: Set<UUID> = []

    /// Cached git status per project.
    var projectGitStatuses: [UUID: GitStatus] = [:]

    /// Cached branch list per project.
    var projectBranches: [UUID: [GitBranch]] = [:]

    /// Projects that are not git repositories.
    var nonGitProjects: Set<UUID> = []

    /// Projects where remote branches could not be fetched.
    var remoteUnavailableProjects: Set<UUID> = []

    /// Branch fetch error messages per project.
    var projectBranchErrors: [UUID: String] = [:]

    /// Cached remote origin URL per project. nil value = no remote configured.
    var projectRemoteURLs: [UUID: String] = [:]

    // MARK: - Branch Operations In-Progress

    /// Keys of the form "\(projectId):\(branchName)" for branches being operated on.
    var branchOperationsInProgress: Set<String> = []

    // MARK: - Commit Panel State (per-project)

    var commitSummaries: [UUID: String] = [:]
    var commitDescriptions: [UUID: String] = [:]
    var generatingAIProjects: Set<UUID> = []
    var committingProjects: Set<UUID> = []
    var commitPanelErrors: [UUID: String] = [:]

    // MARK: - AI Diff Warning Dialog

    var showAIDiffWarning = false
    var pendingAIDiffProject: Project?
    var pendingAIDiffText: String?

    // MARK: - Checkout Error Alert

    /// Unified git operation error (checkout, pull, push).
    var checkoutErrorMessage: String?

    // MARK: - Init

    init(gitService: any GitServicing, aiCommitService: any AICommitServicing) {
        self.gitService = gitService
        self.aiCommitService = aiCommitService
    }

    // MARK: - Git Info Loading

    /// Load git status + branches for a single project.
    /// Each step is independent -- branches failure never blocks the UI.
    func loadGitInfo(for project: Project) async {
        // Step 1: status
        let loadedStatus: GitStatus?
        do {
            let status = try await gitService.status(at: project.path)
            projectGitStatuses[project.id] = status
            nonGitProjects.remove(project.id)
            loadedStatus = status
        } catch let error as GitServiceError {
            if case .notARepository = error {
                nonGitProjects.insert(project.id)
                projectGitStatuses.removeValue(forKey: project.id)
            } else {
                projectBranchErrors[project.id] = error.localizedDescription
            }
            if projectBranches[project.id] == nil {
                projectBranches[project.id] = []
            }
            return
        } catch {
            projectBranchErrors[project.id] = error.localizedDescription
            if projectBranches[project.id] == nil {
                projectBranches[project.id] = []
            }
            return
        }

        // Step 2: branches (independent -- failure shows fallback)
        do {
            let branches = try await gitService.branches(at: project.path)
            projectBranches[project.id] = branches
            remoteUnavailableProjects.remove(project.id)
            projectBranchErrors.removeValue(forKey: project.id)
        } catch {
            remoteUnavailableProjects.insert(project.id)
            if let s = loadedStatus, !s.branch.isEmpty {
                projectBranches[project.id] = [
                    GitBranch(name: s.branch, isRemote: false, isCurrent: true)
                ]
            } else {
                projectBranches[project.id] = []
            }
        }

        // Step 3: remote URL (independent -- nil if not configured)
        await loadRemoteURL(for: project)
    }

    /// Load and cache the remote origin URL for a project.
    func loadRemoteURL(for project: Project) async {
        if let url = await gitService.remoteURL(at: project.path) {
            projectRemoteURLs[project.id] = url
        } else {
            projectRemoteURLs.removeValue(forKey: project.id)
        }
    }

    /// Open the project's remote origin in the default browser.
    func openInRemote(project: Project) {
        guard let rawURL = projectRemoteURLs[project.id],
              let browserURL = GitURLConverter.browserURL(from: rawURL) else {
            return
        }
        NSWorkspace.shared.open(browserURL)
    }

    /// Refresh git status for all projects concurrently (header row badges only).
    func refreshAllGitInfo(projects: [Project]) async {
        let results = await withTaskGroup(
            of: (UUID, Result<GitStatus, Error>).self,
            returning: [(UUID, Result<GitStatus, Error>)].self
        ) { group in
            for project in projects {
                group.addTask { [gitService] in
                    do {
                        let status = try await gitService.status(at: project.path)
                        return (project.id, .success(status))
                    } catch {
                        return (project.id, .failure(error))
                    }
                }
            }

            var collected: [(UUID, Result<GitStatus, Error>)] = []
            for await result in group {
                collected.append(result)
            }
            return collected
        }

        for (projectId, result) in results {
            switch result {
            case .success(let status):
                projectGitStatuses[projectId] = status
                nonGitProjects.remove(projectId)
                projectBranchErrors.removeValue(forKey: projectId)
            case .failure(let error):
                if let gitError = error as? GitServiceError, case .notARepository = gitError {
                    nonGitProjects.insert(projectId)
                    projectGitStatuses.removeValue(forKey: projectId)
                }
            }
        }
    }

    // MARK: - Branch Context Menu Actions

    func gitBranchPull(_ branch: String, isCurrent: Bool, project: Project) async {
        let opKey = "\(project.id):\(branch)"
        branchOperationsInProgress.insert(opKey)
        defer { branchOperationsInProgress.remove(opKey) }

        do {
            let remote = await gitService.defaultRemote(for: branch, at: project.path)
            try await gitService.pullBranch(branch, isCurrent: isCurrent, remote: remote, at: project.path)
            await loadGitInfo(for: project)
        } catch let gitError as GitServiceError {
            switch gitError {
            case .commandFailed(_, _, let stderr):
                checkoutErrorMessage = stderr.trimmingCharacters(in: .whitespacesAndNewlines)
            case .mergeConflict(let files):
                checkoutErrorMessage = "Merge conflict in: \(files.joined(separator: ", "))"
            default:
                checkoutErrorMessage = gitError.localizedDescription
            }
        } catch {
            checkoutErrorMessage = error.localizedDescription
        }
    }

    func gitBranchPush(_ branch: String, project: Project) async {
        let opKey = "\(project.id):\(branch)"
        branchOperationsInProgress.insert(opKey)
        defer { branchOperationsInProgress.remove(opKey) }

        do {
            let remote = await gitService.defaultRemote(for: branch, at: project.path)
            try await gitService.pushBranch(branch, remote: remote, at: project.path)
            await loadGitInfo(for: project)
        } catch let gitError as GitServiceError {
            switch gitError {
            case .commandFailed(_, _, let stderr):
                checkoutErrorMessage = stderr.trimmingCharacters(in: .whitespacesAndNewlines)
            case .pushRejected(let reason):
                checkoutErrorMessage = reason.trimmingCharacters(in: .whitespacesAndNewlines)
            default:
                checkoutErrorMessage = gitError.localizedDescription
            }
        } catch {
            checkoutErrorMessage = error.localizedDescription
        }
    }

    // MARK: - Checkout

    func checkout(branch: String, project: Project) async {
        do {
            try await gitService.checkout(branch: branch, at: project.path)
            await loadGitInfo(for: project)
        } catch let gitError as GitServiceError {
            if case .commandFailed(_, _, let stderr) = gitError {
                checkoutErrorMessage = stderr
                    .trimmingCharacters(in: .whitespacesAndNewlines)
            } else {
                checkoutErrorMessage = gitError.localizedDescription
            }
        } catch {
            checkoutErrorMessage = error.localizedDescription
        }
    }

    // MARK: - Init Repository

    func initRepository(for project: Project) async {
        do {
            try await gitService.initRepository(at: project.path)
            await loadGitInfo(for: project)
        } catch {
            Logger.git.error("git init failed: \(error.localizedDescription, privacy: .public)")
        }
    }

    // MARK: - Commit Actions

    func performCommit(for project: Project) async {
        let summary = (commitSummaries[project.id] ?? "").trimmingCharacters(in: .whitespacesAndNewlines)

        guard !summary.isEmpty else {
            commitPanelErrors[project.id] = "Commit summary cannot be empty"
            return
        }

        let description = (commitDescriptions[project.id] ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        let fullMessage = description.isEmpty ? summary : "\(summary)\n\n\(description)"

        committingProjects.insert(project.id)
        commitPanelErrors.removeValue(forKey: project.id)
        defer { committingProjects.remove(project.id) }

        do {
            try await gitService.stage(files: [], at: project.path)
            try await gitService.commit(message: fullMessage, at: project.path)
            commitSummaries.removeValue(forKey: project.id)
            commitDescriptions.removeValue(forKey: project.id)
            await loadGitInfo(for: project)
        } catch let gitError as GitServiceError {
            if case .commandFailed(_, _, let stderr) = gitError {
                commitPanelErrors[project.id] = stderr.trimmingCharacters(in: .whitespacesAndNewlines)
            } else {
                commitPanelErrors[project.id] = gitError.localizedDescription
            }
        } catch {
            commitPanelErrors[project.id] = error.localizedDescription
        }
    }

    /// First step of AI commit: get diff and show confirmation dialog.
    func generateAICommitMessage(for project: Project) async {
        commitPanelErrors.removeValue(forKey: project.id)

        do {
            let diff = try await gitService.headDiff(at: project.path)
            guard !diff.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
                commitPanelErrors[project.id] = "No changes to analyze"
                return
            }

            pendingAIDiffText = diff
            pendingAIDiffProject = project
            showAIDiffWarning = true
        } catch {
            commitPanelErrors[project.id] = "AI: \(error.localizedDescription)"
        }
    }

    /// Called after the user confirms the AI diff warning dialog.
    func sendAIDiff(_ diff: String, for project: Project) async {
        generatingAIProjects.insert(project.id)
        commitPanelErrors.removeValue(forKey: project.id)
        defer { generatingAIProjects.remove(project.id) }

        do {
            let truncated = String(diff.prefix(AIConstants.maxDiffLength))
            let result = try await aiCommitService.generateCommitMessage(for: truncated)

            let trimmed = result.trimmingCharacters(in: .whitespacesAndNewlines)
            let lines = trimmed.components(separatedBy: "\n")
            let summaryLine = lines.first?.trimmingCharacters(in: .whitespaces) ?? ""
            let descPart = lines.dropFirst()
                .drop(while: { $0.trimmingCharacters(in: .whitespaces).isEmpty })
            let descText = Array(descPart).joined(separator: "\n")
                .trimmingCharacters(in: .whitespacesAndNewlines)

            commitSummaries[project.id] = summaryLine
            commitDescriptions[project.id] = descText
        } catch {
            commitPanelErrors[project.id] = "AI: \(error.localizedDescription)"
        }
    }
}

File: shared/src/commonMain/kotlin/studio/vibe/shared/viewmodel/GitSidebarViewModel.kt
package studio.vibe.shared.viewmodel

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import studio.vibe.shared.contract.AICommitServicing
import studio.vibe.shared.contract.GitServicing
import studio.vibe.shared.contract.UrlConverting
import studio.vibe.shared.contract.UrlOpening
import studio.vibe.shared.model.FilePath
import studio.vibe.shared.model.GitBranch
import studio.vibe.shared.model.GitStatus
import studio.vibe.shared.service.git.GitURLConverter
import studio.vibe.shared.viewmodel.git.BranchActionsHandler
import studio.vibe.shared.viewmodel.git.CommitActionsHandler
import studio.vibe.shared.viewmodel.git.RemoteActionsHandler
import kotlin.uuid.ExperimentalUuidApi
import kotlin.uuid.Uuid

@OptIn(ExperimentalUuidApi::class)
data class GitSidebarState(
    val gitExpandedProjects: Set<Uuid> = emptySet(),
    val projectGitStatuses: Map<Uuid, GitStatus> = emptyMap(),
    val projectBranches: Map<Uuid, List<GitBranch>> = emptyMap(),
    val nonGitProjects: Set<Uuid> = emptySet(),
    val remoteUnavailableProjects: Set<Uuid> = emptySet(),
    val commitSummaries: Map<Uuid, String> = emptyMap(),
    val commitDescriptions: Map<Uuid, String> = emptyMap(),
    val commitPanelErrors: Map<Uuid, String> = emptyMap(),
    val generatingAIProjects: Set<Uuid> = emptySet(),
    val committingProjects: Set<Uuid> = emptySet(),
    val checkoutErrorMessage: String? = null,
    /**
     * Branch name pending delete confirmation per project.
     * Non-null means a confirmation dialog should be shown.
     * Cleared after user confirms or cancels.
     */
    val pendingDeleteBranch: Map<Uuid, String> = emptyMap(),
    val deleteErrorMessage: String? = null,
)

/**
 * Thin orchestrator: owns [GitSidebarState] and delegates all actions to focused handlers.
 *
 * Public API is identical to the previous monolithic implementation — all callers
 * (UI, DI, tests) are unaffected.
 */
@OptIn(ExperimentalUuidApi::class)
class GitSidebarViewModel(
    private val gitService: GitServicing,
    private val aiCommitService: AICommitServicing,
    parentScope: CoroutineScope,
    private val urlOpening: UrlOpening? = null,
    private val urlConverter: UrlConverting = GitURLConverter,
) : BaseViewModel(parentScope) {

    private val _state = MutableStateFlow(GitSidebarState())
    val state: StateFlow<GitSidebarState> = _state.asStateFlow()

    // ── Action handlers ────────────────────────────────────────────────────────

    private val branchActions = BranchActionsHandler(
        gitService = gitService,
        scope = scope,
        state = _state,
        loadGitInfo = ::loadGitInfo,
    )

    private val commitActions = CommitActionsHandler(
        gitService = gitService,
        aiCommitService = aiCommitService,
        scope = scope,
        state = _state,
        loadGitInfo = ::loadGitInfo,
    )

    private val remoteActions = RemoteActionsHandler(
        gitService = gitService,
        scope = scope,
        state = _state,
        loadGitInfo = ::loadGitInfo,
        urlOpening = urlOpening,
        urlConverter = urlConverter,
    )

    // ── State helpers ──────────────────────────────────────────────────────────

    fun toggleExpanded(projectId: Uuid) {
        _state.update { s ->
            val expanded = s.gitExpandedProjects.toMutableSet()
            if (projectId in expanded) expanded.remove(projectId) else expanded.add(projectId)
            s.copy(gitExpandedProjects = expanded)
        }
    }

    fun loadGitInfo(projectId: Uuid, path: FilePath) {
        scope.launch {
            runCatching {
                val isRepo = gitService.isRepository(at = path)
                if (!isRepo) {
                    _state.update { s ->
                        s.copy(nonGitProjects = s.nonGitProjects + projectId)
                    }
                    return@launch
                }
                val status = gitService.status(at = path)
                val branches = gitService.branches(at = path)
                _state.update { s ->
                    s.copy(
                        projectGitStatuses = s.projectGitStatuses + (projectId to status),
                        projectBranches = s.projectBranches + (projectId to branches),
                        nonGitProjects = s.nonGitProjects - projectId,
                    )
                }
                // Check remote availability
                runCatching {
                    gitService.remoteURL(name = "origin", at = path)
                }.onFailure {
                    _state.update { s ->
                        s.copy(remoteUnavailableProjects = s.remoteUnavailableProjects + projectId)
                    }
                }.onSuccess { url ->
                    if (url == null) {
                        _state.update { s ->
                            s.copy(remoteUnavailableProjects = s.remoteUnavailableProjects + projectId)
                        }
                    } else {
                        _state.update { s ->
                            s.copy(remoteUnavailableProjects = s.remoteUnavailableProjects - projectId)
                        }
                    }
                }
            }.onFailure {
                _state.update { s ->
                    s.copy(nonGitProjects = s.nonGitProjects + projectId)
                }
            }
        }
    }

    fun refreshAllGitInfo(projects: List<Pair<Uuid, FilePath>>) {
        projects.forEach { (id, path) -> loadGitInfo(id, path) }
    }

    fun cleanupProject(projectId: Uuid) {
        _state.update { s ->
            s.copy(
                gitExpandedProjects = s.gitExpandedProjects - projectId,
                projectGitStatuses = s.projectGitStatuses - projectId,
                projectBranches = s.projectBranches - projectId,
                nonGitProjects = s.nonGitProjects - projectId,
                remoteUnavailableProjects = s.remoteUnavailableProjects - projectId,
                commitSummaries = s.commitSummaries - projectId,
                commitDescriptions = s.commitDescriptions - projectId,
                commitPanelErrors = s.commitPanelErrors - projectId,
                generatingAIProjects = s.generatingAIProjects - projectId,
                committingProjects = s.committingProjects - projectId,
                pendingDeleteBranch = s.pendingDeleteBranch - projectId,
            )
        }
    }

    // ── Branch delegation ──────────────────────────────────────────────────────

    fun checkout(projectId: Uuid, branch: String, path: FilePath) =
        branchActions.checkout(projectId, branch, path)

    fun checkoutBranch(projectId: Uuid, branchName: String, path: FilePath) =
        branchActions.checkout(projectId, branchName, path)

    fun requestDeleteBranch(projectId: Uuid, branchName: String) =
        branchActions.requestDeleteBranch(projectId, branchName)

    fun cancelDeleteBranch(projectId: Uuid) =
        branchActions.cancelDeleteBranch(projectId)

    fun confirmDeleteBranch(projectId: Uuid, path: FilePath, force: Boolean = false) =
        branchActions.confirmDeleteBranch(projectId, path, force)

    fun clearCheckoutError() =
        branchActions.clearCheckoutError()

    fun clearDeleteError() =
        branchActions.clearDeleteError()

    // ── Commit delegation ──────────────────────────────────────────────────────

    fun performCommit(projectId: Uuid, path: FilePath) =
        commitActions.performCommit(projectId, path)

    fun generateAICommitMessage(projectId: Uuid, path: FilePath) =
        commitActions.generateAICommitMessage(projectId, path)

    fun sendAIDiff(projectId: Uuid, path: FilePath) =
        commitActions.generateAICommitMessage(projectId, path)

    fun updateCommitSummary(projectId: Uuid, value: String) =
        commitActions.updateCommitSummary(projectId, value)

    fun updateCommitDescription(projectId: Uuid, value: String) =
        commitActions.updateCommitDescription(projectId, value)

    // ── Remote delegation ──────────────────────────────────────────────────────

    fun gitBranchPull(projectId: Uuid, branch: String, isCurrent: Boolean, path: FilePath) =
        remoteActions.gitBranchPull(projectId, branch, isCurrent, path)

    fun gitBranchPush(projectId: Uuid, branch: String, path: FilePath) =
        remoteActions.gitBranchPush(projectId, branch, path)

    fun openInRemote(projectId: Uuid, path: FilePath) =
        remoteActions.openInRemote(projectId, path)

    // ── Repository init ────────────────────────────────────────────────────────

    fun initRepository(projectId: Uuid, path: FilePath) {
        scope.launch {
            runCatching {
                gitService.initRepository(at = path)
                loadGitInfo(projectId, path)
            }.onFailure { e ->
                _state.update { s ->
                    s.copy(
                        commitPanelErrors = s.commitPanelErrors + (projectId to (e.message ?: "Init failed"))
                    )
                }
            }
        }
    }
}
